### PR TITLE
refactor: migrate to ESM

### LINKED ISSUE #7330: [Architecture] Switch package.json to module
#### Summary

After trying to implement #7329, Jest's ESM requires that every module file be moved to `.mjs` or switch the package to a module. So, moving to Vite/Jest may be split into three different PRs, with the first one switching this repo to a module (which _should_ help with bundle sizes). Using Karma + ESM relies on a ([small CJS workaround](https://github.com/karma-runner/karma/issues/3677#issuecomment-1460235822)). This should also unblock #6591 and #6592

### PR DESCRIPTION
**Pull request by LeoDog896**
Closes 

### Describe your changes:
Fixes #7330 (and continues the spirit of https://github.com/nasa/openmct/tree/esm), to hopefully unblock a few pending issues. I'm opening this now to test its effects on CI as I work on this. Import transformations were done via codemods 👍 

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing document](https://github.com/nasa/openmct/blob/master/CONTRIBUTING.md)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/nasa/openmct/pulls) for the same update/change?
* [x] Is this change backwards compatible? For example, developers won't need to change how they are calling the API or how they've extended core plugins such as Tables or Plots.

### Author Checklist

* [x] Changes address original issue?
* [ ] Tests included and/or updated with changes?
* [ ] Has this been smoke tested?
* [x] Have you associated this PR with a `type:` label? Note: this is not necessarily the same as the original issue.
* [ ] Have you associated a milestone with this PR? Note: leave blank if unsure.
* [ ] Is this a breaking change to be called out in the release notes?
* [ ] Testing instructions included in associated issue OR is this a dependency/testcase change?

### Reviewer Checklist

* [ ] Changes appear to address issue?
* [ ] Reviewer has tested changes by following the provided instructions?
* [ ] Changes appear not to be breaking changes?
* [ ] Appropriate automated tests included?
* [ ] Code style and in-line documentation are appropriate?

### CODE REVIEW

File: e2e/helper/notebookUtils.js [hunk 1]
```diff
@@ -20,11 +20,12 @@
  * at runtime from the About dialog for additional information.
  *****************************************************************************/
 
-const { createDomainObjectWithDefaults } = require('../appActions');
+import { createDomainObjectWithDefaults } from '../appActions.js';
 
 const NOTEBOOK_DROP_AREA = '.c-notebook__drag-area';
 const CUSTOM_NAME = 'CUSTOM_NAME';
-const path = require('path');
+import path from 'path';
```
[the remark this answers is not in the document]
Comment (LeoDog896) [reply e2e/helper/notebookUtils.js]: 👍

File: e2e/helper/faultUtils.js [hunk 2]
```diff
@@ -19,14 +19,16 @@
  * this source code distribution or the Licensing information page available
  * at runtime from the About dialog for additional information.
  *****************************************************************************/
-/* global __dirname */
-const path = require('path');
+import path from 'path';
```
[the remark this answers is not in the document]
Comment (LeoDog896) [reply e2e/helper/faultUtils.js]: 👍

**Comment by LeoDog896** [comment]
I have no clue why that test is failing - and it's definitely not flaky, I can reproduce the same behavior locally. I probably made a mistake when converting.

File: karma.conf.cjs [hunk 3]
```diff
@@ -22,17 +22,17 @@
...
 
-module.exports = (config) => {
+module.exports = async (config) => {
   let webpackConfig;
   let browsers;
   let singleRun;
 
   if (process.env.KARMA_DEBUG) {
-    webpackConfig = require('./.webpack/webpack.dev.js');
+    webpackConfig = (await import('./.webpack/webpack.dev.js')).default;
```
Comment (unlikelyzero) [inline karma.conf.cjs:31]: interesting

File: src/plugins/plugins.js [hunk 4]
```diff
@@ -20,70 +20,70 @@
...
+import Timeline from './timeline/plugin.js';
+import TimeList from './timelist/plugin.js';
+import Timer from './timer/plugin.js';
+import URLIndicatorPlugin from './URLIndicatorPlugin/URLIndicatorPlugin.js';
+import URLTimeSettingsSynchronizer from './URLTimeSettingsSynchronizer/plugin.js';
+import UserIndicator from './userIndicator/plugin.js';
+import UTCTimeSystem from './utcTimeSystem/plugin.js';
+import ViewDatumAction from './viewDatumAction/plugin.js';
+import ViewLargeAction from './viewLargeAction/plugin.js';
+import WebPagePlugin from './webPage/plugin.js';
```
Comment (unlikelyzero) [inline src/plugins/plugins.js:86]: i'm  just going to assume these are all correct 😅

**Review by unlikelyzero — APPROVED** [review]
This looks great. This is amazing work. In terms of review, everything looks correct.

We will need the impact this will have to projects which treat open mct as a npm dependency and also build it from source (like akhenry/openmct-yamcs)

File: karma.conf.cjs (hunk 3 above)
Comment (LeoDog896) [reply karma.conf.cjs:31]: yeah, CJS can't require an ESM module, but it _can_ use `await import`.

File: src/plugins/plugins.js (hunk 4 above)
Comment (LeoDog896) [reply src/plugins/plugins.js:86]: the diff helps 👍 - those `.js` transformations were done with regex

**Comment by LeoDog896** [comment]
> We will need the impact this will have to projects which treat open mct as a npm dependency and also build it from source (like akhenry/openmct-yamcs)

I didn't quite think of that - it does modify the `openmct.js` entry file, so this is technically a breaking change if they use that file

**Comment by ozyx** [comment]
I'll go ahead and do some cursory testing against some of the consuming projects.

File: e2e/README.md [hunk 5]
```diff
@@ -480,7 +480,7 @@ The following contains a list of tips and tricks which don't exactly fit into a
 It is possible to override the browser's clock in order to control time-based elements. Since this can cause unwanted behavior (i.e. Tree not rendering), only use this sparingly. To do this, use the `overrideClock` fixture as such:
 
 ```js
-const { test, expect } = require('../../pluginFixtures.js');
+import { test, expect } from '../../pluginFixtures.js';
```
Comment (ozyx) [inline e2e/README.md:483]: Nice! :)

File: e2e/appActions.js [hunk 6]
```diff
@@ -644,25 +644,24 @@ async function renameObjectFromContextMenu(page, url, newName) {
   await page.click('[aria-label="Save"]');
 }
 
-// eslint-disable-next-line no-undef
-module.exports = {
+export {
```
Comment (ozyx) [inline e2e/appActions.js:647]: Lovely

**Review by ozyx — APPROVED** [review]
Everything looks great! Tested this against some projects that use Open MCT as a dependency and everything just works. Thank you so much for taking this on! 🚀🚀🚀

**Pull request merged**